helix-term/src/handlers/completion.rs

```diff
@@ -291,6 +301,226 @@ fn request_completion(
+                    } else if matched_path.starts_with("${HOME}") {
+                        Some(7)
+                    } else {
+                        None
+                    };
+                    if let Some(start_char) = first_separator_after_home {
+                        let mut path = home.to_owned();
+                        path.push(&matched_path[start_char..]);
+                        path
+                    } else {
+                        matched_path.into()
+                    }
+                } else {
+                    matched_path.into()
+                };
+                PathBuf::from(home_resolved_path)
+            };
+            #[cfg(not(unix))]
+            let mut path = PathBuf::from(matched_path);
+
+            if path.is_relative() {
+                if let Some(doc_path) = doc.path().and_then(|dp| dp.parent()) {
+                    path = doc_path.join(path);
+                } else if let Ok(work_dir) = std::env::current_dir() {
+                    path = work_dir.join(path);
+                }
+            }
+            let ends_with_slash = match matched_path.chars().last() {
+                Some(std::path::MAIN_SEPARATOR) => true,
+                None => return None,
+                _ => false,
+            };
+            // check if there are chars after the last slash, and if these chars represent a directory
+            match std::fs::metadata(path.clone()).ok() {
+                Some(m) if m.is_dir() && ends_with_slash => {
+                    Some((PathBuf::from(path.as_path()), None))
+                }
+                _ if !ends_with_slash => path.parent().map(|parent_path| {
+                    (
+                        PathBuf::from(parent_path),
+                        path.file_name().and_then(|f| f.to_str().map(String::from)),
+                    )
+                }),
+                _ => None,
+            }
+        })?;
+
+    // The async file accessor functions of tokio were considered, but they were a bit slower
+    // and less ergonomic than just using the std functions in a separate "thread"
+    let future = tokio::task::spawn_blocking(move || {
+        let Some(read_dir) = std::fs::read_dir(&dir_path).ok() else {
+            return Vec::new();
+        };
+
+        read_dir
+            .filter_map(Result::ok)
+            .filter_map(|dir_entry| {
+                let path = dir_entry.path();
+                // check if <chars> in <path>/<chars><cursor> matches the start of the filename
+                let filename_starts_with_prefix =
+                    match (path.file_name().and_then(OsStr::to_str), &typed_file_name) {
+                        (Some(re_stem), Some(t)) => re_stem.starts_with(t),
+                        _ => true,
+                    };
+                if filename_starts_with_prefix {
+                    dir_entry.metadata().ok().map(|md| (path, md))
+                } else {
+                    None
+                }
```

Comment:
I believe we want to skip the prefix check here and delegate the filtering to the ui::Menu and Nucleo. This makes a difference if you trigger completion manually, for example if you're in the helix repo and type `./lsp` you'll see `helix-lsp` and `helix-lsp-types`. If you then hit `C-c` and `C-x` to manually trigger completion though you won't have any suggestions because of the prefix check.

So instead this could be `.filter_map(|dir_entry| dir_entry.metadata().ok().map(|md| (dir_entry.path(), md)))`

Reply:
Yeah makes sense, it also allows case-insensitive completion I think.